numpy/_array_api/_array_object.py

```diff
@@ -0,0 +1,986 @@
+                    if len(key) == 1:
+                        return key
+                    raise IndexError("Boolean array indices combined with other indices are not allowed in the array API namespace")
+                if isinstance(idx, tuple):
+                    raise IndexError("Nested tuple indices are not allowed in the array API namespace")
+
+            if shape is None:
+                return key
+            n_ellipsis = key.count(...)
+            if n_ellipsis > 1:
+                return key
+            ellipsis_i = key.index(...) if n_ellipsis else len(key)
+
+            for idx, size in list(zip(key[:ellipsis_i], shape)) + list(zip(key[:ellipsis_i:-1], shape[:ellipsis_i:-1])):
+                Array._validate_index(idx, (size,))
+            return key
+        elif isinstance(key, bool):
+            return key
+        elif isinstance(key, Array):
+            if key.dtype in _integer_dtypes:
+                if key.ndim != 0:
+                    raise IndexError("Non-zero dimensional integer array indices are not allowed in the array API namespace")
+            return key._array
+        elif key is Ellipsis:
+            return key
+        elif key is None:
+            raise IndexError("newaxis indices are not allowed in the array API namespace")
+        try:
+            return operator.index(key)
+        except TypeError:
+            # Note: This also omits boolean arrays that are not already in
+            # Array() form, like a list of booleans.
+            raise IndexError("Only integers, slices (`:`), ellipsis (`...`), and boolean arrays are valid indices in the array API namespace")
+
+    # Everything below this line is required by the spec.
+
+    def __abs__(self: Array, /) -> Array:
+        """
+        Performs the operation __abs__.
+        """
+        res = self._array.__abs__()
+        return self.__class__._new(res)
+
+    def __add__(self: Array, other: Union[int, float, Array], /) -> Array:
+        """
+        Performs the operation __add__.
+        """
+        other = self._check_allowed_dtypes(other, 'numeric', '__add__')
+        if other is NotImplemented:
+            return other
+        self, other = self._normalize_two_args(self, other)
+        res = self._array.__add__(other._array)
+        return self.__class__._new(res)
+
+    def __and__(self: Array, other: Union[int, bool, Array], /) -> Array:
+        """
+        Performs the operation __and__.
+        """
+        other = self._check_allowed_dtypes(other, 'integer or boolean', '__and__')
+        if other is NotImplemented:
+            return other
+        self, other = self._normalize_two_args(self, other)
+        res = self._array.__and__(other._array)
+        return self.__class__._new(res)
+
+    def __array_namespace__(self: Array, /, *, api_version: Optional[str] = None) -> object:
+        if api_version is not None:
```

Comment:
@rgommers what spec version should we accept here as the version implemented by numpy._array_api? I couldn't tell what version the spec is at, if it does currently have one.

Reply:
Good question. The version scheme is at https://data-apis.org/array-api/latest/future_API_evolution.html#versioning. But since we're not at the first stable release, the current `conf.py` of the Sphinx site simply uses:
```
release = '2021.01-DRAFT' 
```
I'd say in this `if api_version` check accept anything that starts with `'2021`, if it says `2022` or something random, that'd be user error or a future version that's not yet supported.